README.md
# Paperboy

A wrapper around AIOKafka to make it easier to use, for message ingestion in your micro services

[[_TOC_]]

## Installation

```bash
pip install kafka-paperboy
```

## Quick Start

```python
import asyncio
from paperboy import Engine, Handler

# Create a topic handler
class ExampleTopicHandler(Handler):
    topic = "example-topic"

    @classmethod
    async def on_message(cls, msg, ctx):
        print(f"Received message: [{msg.key}] {msg.value}")

# Create an instance of the consumer engine and start it
c = Engine(
    ExampleTopicHandler,
    bootstrap_servers="localhost:9092",
    group_id="example-group",
    fail_on_exception=True
)
asyncio.run(c.start())
```

## The Engine

### Main idea

The engine here is a wrapper around the AIOKafkaConsumer, that will handle the consumption of messages from Kafka, and dispatch them to the right handler, based on the topic.

Rather than defining all the consumption logic in a single place, the engine will dispatch to handlers, that you can parameter by coding their lifecycle methods.

This engine is meant to be used in a micro service architecture, where you have multiple services, each one consuming messages from a set of topics, and doing some processing on them.

3 modes are available:

- **Single mode**: The engine will consume messages from Kafka, and dispatch them to the handlers, message per message. This is the default mode
- **Bulk mode**: The engine will consume messages from Kafka by batch (based on the number of records the engine can take, or a timeout), and dispatch them to the handlers, in a synchronous way. This mode is useful if you want to do some processing on the messages, and then commit the offsets, to avoid losing messages in case of failure.
- **Step mode**: The engine will consume certain topics from Kafka, and dispatch them to the handlers, in bulkd. This mode is useful if you want to consume certain topics, before others. After finishing the last steps, the engine will go back to single mode with all the topics handlers.

Single mode is the default mode, but the engine will switch to Bulk mode if the lag of the consumer is too high, to avoid overloading the Kafka cluster. Using Bulk mode is recommended if you want to do some processing on the messages, and then commit the offsets, to avoid losing messages in case of failure.
Once the lag of the consumer is back to normal, the engine will switch back to Single mode, to keep the reactivity of the service.

### Step mode

The engine can be run in Step mode, to consume certain topics before others. This is useful if you want to consume certain topics, before others. After finishing the last steps, the engine will go back to single mode with all the topics handlers.

To enable this mode, you need to define the `steps` attribute of the engine, as a list of lists of topics. Each list of topics will be consumed in bulk, and the engine will go back to single mode after finishing the last step.

```python
import asyncio
from paperboy import Engine

c = Engine(
    ExampleOneHandler,
    ExampleTwoHandler,
    ExampleThreeHandler,
    steps = [
        [ExampleOneHandler, ExampleTwoHandler,],
        [ExampleThreeHandler],
    ]
    bootstrap_servers="localhost:9092",
    group_id="example-group",
    fail_on_exception=True
)
asyncio.run(c.start())
```

## Handlers

Handlers are the classes that will handle the messages from a Kafka cluster.
Those handlers treats a series of messages, based on the topic they are subscribed to.
A handler exposes a series of class methods, used to handle the messages on certain points of their lifecycle, meant to be overriden by the user.

### Handler Quick Start

```python
import asyncio
from paperboy import Handler, Engine

class ExampleHandler(Handler):
    topic = "example-topic"

    @classmethod
    async def on_message(cls, msg, ctx):
        # Lifecycle method called when a message is received
        print(f"Received message: [{msg.key}] {msg.value}")

    @classmethod
    async def on_error(cls, e, ctx, exc) -> Exception | None:
        # Lifecycle method called when an error is raised in the handler
        print(f"Handler error: {exc}")
        return e

c = Engine(
    ExampleHandler,
    bootstrap_servers="localhost:9092",
    group_id="example-group",
    auto_offset_reset="earliest",
    enable_auto_commit=False,
)
asyncio.run(c.start())
```

### Handler class

You can define the following class attributes on a handler:
    - **topic**: The topic to consume
    - **key_serializer**: The serializer for the key of the message
    - **value_serializer**: The serializer for the value of the message

Serializers must be a callable, that takes a bytes object as input, and returns a deserialized object.

```python
class ExampleHandler(Handler):
    topic: str = "example-topic" # Topic to consume
    key_serializer: Callable = lambda x: json.loads(x.decode("utf-8")) # Serializer for the key of the message
    value_serializer: Callable = lambda x: json.loads(x.decode("utf-8")) # Serializer for the value of the message
```

This works alors with Avro schemas, using the `python-schema-registry-client` library

```python
from schema_registry.client import SchemaRegistryClient
from schema_registry.serializers.avro import AvroMessageSerializer

client = SchemaRegistryClient(url="http://localhost:8081")
serializer = AvroMessageSerializer(client)

class ExampleHandler(Handler):
    topic: str = "example-topic" # Topic to consume
    key_serializer: Callable = serializer.decode_message # Serializer for the key of the message
    value_serializer: Callable = serializer.decode_message # Serializer for the value of the message
```

### Handler lifecycle methods

#### on_message

```python
    @classmethod
    async def on_message(cls, msg: ConsumerRecord, ctx: Context):
        pass
```

This method is called when a message is received from Kafka. It takes 2 arguments:

- **msg**: The message received from Kafka, as a ConsumerRecord (from aiokafka)
- **ctx**: The context of the message, as a Context object (defined in the define_context method)

Returns nothing

#### on_tombstone

```python
    @classmethod
    async def on_tombstone(cls, msg: ConsumerRecord, ctx: Context):
        pass
```

This method is called when a tombstone message is received from Kafka.
Tombstone messages are messages with a null value, used to delete a key from a compacted topic.

It takes 2 arguments:

- **msg**: The message received from Kafka, as a ConsumerRecord (from aiokafka)
- **ctx**: The context of the message, as a Context object (defined in the define_context method)

Returns nothing

#### on_error

```python
    @classmethod
    async def on_error(cls, e: Exception, msg: ConsumerRecord, ctx: Context) -> Exception | None:
        pass
```

This method is called when an error is raised in the handler.

It takes 3 arguments:

- **e**: The exception raised in the handler
- **msg**: The message received from Kafka, as a ConsumerRecord (from aiokafka)
- **ctx**: The context of the message, as a Context object (defined in the define_context method)

Returns an Exception, or None. If an Exception is returned, it will be raised in the engine, and the engine will stop. If None is returned, the engine will log the exception, and continue the consumption of messages.

#### did_receive_message

```python
    @classmethod
    async def did_receive_message(cls, msg: ConsumerRecord, ctx: Context) -> ConsumerRecord:
        return msg
```

This method is called after the message reception.
This method is useful if you want to do some processing on the message, before it is handled by the on_message method.

It takes 2 arguments:

- **msg**: The message received from Kafka, as a ConsumerRecord (from aiokafka)
- **ctx**: The context of the message, as a Context object (defined in the define_context method)

Returns a ConsumerRecord.

#### on_finished_handling

```python
    @classmethod
    async def on_finished_handling(cls, msgs: ConsumerRecord | list[ConsumerRecord], ctx: Context):
        pass
```

This method is called after the overall message handling.
This method is useful if you want to do some processing on the message, after it has been handled by the on_message / on_tombstone method.

It takes 2 arguments:

- **msgs**: The message received from Kafka, as a ConsumerRecord (from aiokafka), or a list of ConsumerRecord if the engine is in bulk mode
- **ctx**: The context of the message, as a Context object (defined in the define_context method)

Returns nothing.

#### define_context

```python
    @classmethod
    async def define_context(cls) -> Context | None:
        return {}
```

This method is called before the message handling, and is used to define the context of the handling.
The context is a dictionary, that will be passed to the lifecycle methods, and can be used to store some data, that will be used in the lifecycle methods.

It takes no arguments.

Returns a dictionary, or None.

### Bulk Handlers

Bulk handlers are handlers that will handle a batch of messages, rather than a single message.
Those handlers are usefull if you want to apply a specific logic on a batch of messages, rather than a single message, via dedicated lifecycle methods.

You also need to specify normal lifecycle methods with the bulk ones, to handle the messages when the engine runs in Single Mode

```python
import asyncio
from paperboy import BulkHandler, Engine

class ExampleProbeHandler(BulkHandler):

    @classmethod
    async def on_bulk(cls, msgs: list[ConsumerRecord], ctx: Context):
        # Lifecycle method called when a batch of messages is received
        print(f"Received {len(msgs)} messages")

    @classmethod
    async def on_message(cls, msg: ConsumerRecord, ctx: Context):
        # Lifecycle method called when a message is received
        print(f"Received message: [{msg.key}] {msg.value}")


c = KafkaConsumerEngine(
    ExampleHandler,
    bootstrap_servers="localhost:9092",
    with_bulk_mode=True
    group_id="example-group",
    auto_offset_reset="earliest",
    enable_auto_commit=False,
)
asyncio.run(c.start())
```

In this example, the engine will run in Bulk Mode, and will call the on_bulk_message method, when a batch of messages is received, rather than applying the on_message method on each message. The on_message method is still needed, to handle the messages when the engine runs in Single Mode.

### BulkHandler parameters

#### on_bulk

```python
    @classmethod
    async def on_bulk(cls, msgs: list[ConsumerRecord], ctx: Context):
        pass
```

This method is called when a batch of messages is received from Kafka.

It takes 2 arguments:

- **msgs**: The messages received from Kafka, as a list of ConsumerRecord (from aiokafka)
- **ctx**: The context of the message, as a Context object (defined in the define_context method)

Returns nothing

#### on_bulk_error

```python
    @classmethod
    async def on_bulk_error(cls, e: Exception, msgs: list[ConsumerRecord], ctx: Context) -> Exception | None:
        pass
```

This method is called when an error is raised in the handler, when the engine is in Bulk Mode.

It takes 3 arguments:

- **e**: The exception raised in the handler
- **msgs**: The messages received from Kafka, as a list of ConsumerRecord (from aiokafka)
- **ctx**: The context of the message, as a Context object (defined in the define_context method)

Returns an Exception, or None. If an Exception is returned, it will be raised in the engine, and the engine will stop. If None is returned, the engine will log the exception, and continue the consumption of messages.

#### did_receive_bulk_messages

```python
    @classmethod
    async def did_receive_bulk_messages(cls, msgs: list[ConsumerRecord], ctx: Context) -> list[ConsumerRecord]:
        return msgs
```

This method is called after the batch of messages reception. This method is useful if you want to do some processing on the batch of messages, before it is handled by the on_bulk method.

It takes 2 arguments:

- **msgs**: The messages received from Kafka, as a list of ConsumerRecord (from aiokafka)
- **ctx**: The context of the message, as a Context object (defined in the define_context method)

Returns a list of ConsumerRecord.

## Tables

Tables are key/value stores, that can be used to store computed data from the messages, and can be used in the handlers.
To ensure Table consistency, each mutation of the table is stored in a compacted topic in Kafka, and the table is reloaded from the topic at the start of the engine.

### Table Quick Start

```python
import asyncio
from paperboy import Table, Engine

metrics_table = Table(
    "metrics",
    store="memory://"
)

class PageViewHandler(Handler):
    topic = "page-views"

    @classmethod
    async def on_message(cls, msg, ctx):
        # Access a specific key of the table (should return None if the key is not set)
        clicks = metrics_table.get("page_views")
        # Set a new value
        await metrics_table.set("page_views", (clicks or 0) + 1)

    @classmethod
    async def on_tombstone(cls, msg, ctx):
        # Delete a key
        await metrics_table.delete("page_views")

engine = Engine(
    PageViewHandler,
    bootstrap_servers="localhost:9092",
    group_id="example-group",
    fail_on_exception=True
)

# Register the table in the engine (the table will be reloaded from the topic at the start of the engine)
engine.register_table(metrics_table)

asyncio.run(engine.start())
```

### Table stores

Tables uses 2 types of backend to store the data:

- **InMemory**: The data is stored in memory, and is lost when the engine stops
- **RocksDB**: The data is stored in a RocksDB database, and is persisted in the worker

InMemory database is preferred in development, but RocksDB is recommended in production, to keep the data in case of failure.

```python
from paperboy import Table

# InMemory database
metrics_table = Table("metrics", store="memory://")

# RocksDB database
persisted_metrics_table = Table("prod_metrics", store="rocksdb://")
```

### Table methods

```python
from paperboy import Table

table = Table(
    # Table name
    "metrics",
    # Table store
    store="memory://" if DEBUG else "rocksdb://",
    # Changelog topic name in Kafka Cluster
    changelog_topic_name="metrics-changelog",
    # Changelog topic configuration
    changelog_topic_num_partitions = 1,
    changelog_topic_replication_factor = 1,
    changelog_topic_segment_ms = 43200000,
    changelog_topic_min_cleanable_dirty_ratio = 0.01,
    changelog_topic_max_compaction_lag_ms = 86400000,        
)

# Get a value from the table (returns None if the key is not set)
value = table.get("key") 

# Set a value in the table
await table.set("key", "value")

# Delete a value from the table
await table.delete("key")
```

## Healthcheck

The engine exposes a healthcheck endpoint, that can be used to check if the engine still ingest.

It's exposed on the 8080 port by default and can be configured with 2 arguments:

- **healthcheck_port**: The port to expose the healthcheck endpoint (8080 by default)
- **healthcheck_host**: The host to expose the healthcheck endpoint ("0.0.0.0" by default)

```python
from paperboy import Engine

engine = Engine(
    PageViewHandler,
    bootstrap_servers="localhost:9092",
    group_id="example-group",
    fail_on_exception=True
    healthcheck_port=8080,
    healthcheck_host="0.0.0.0"
)

asyncio.run(engine.start())
```


CHANGELOG.md
## 0.5.0 (2023-12-29)

### Feat

- add minimal healthcheck API

## 0.4.1 (2023-12-29)

### Fix

- change fail_on_exception logic

## 0.4.0 (2023-12-29)

### Feat

- add tombstone and values split on batch helper

### Fix

- put preflight hooks in try block for bulk handler
- change get_last_values_from_batch logic to properly check last values

## 0.3.0 (2023-12-04)

### Feat

- add tables feature

### Fix

- **engine**: seek partitions on single mode when out of bulk mode on no_commit

## 0.2.0 (2023-11-27)

## 0.1.0 (2023-11-27)

### Fix

- add logger as class attribute


paperboy/engine.py
import asyncio
import logging
import logging.config
import signal
from typing import final

from aiokafka import AIOKafkaConsumer, AIOKafkaProducer, ConsumerRebalanceListener, ConsumerStoppedError, TopicPartition
from aiokafka.helpers import create_ssl_context

from paperboy.handler import BaseHandler
from paperboy.health_api import run_api
from paperboy.table import Table

from .logs import PaperboyFormatter


class EngineConsumerRebalancer(ConsumerRebalanceListener):
    def __init__(self, lock):
        self.lock = lock

        self.log = logging.getLogger("paperboy.ConsumerRebalancer")
        ch = logging.StreamHandler()
        ch.setLevel(logging.INFO)
        ch.setFormatter(PaperboyFormatter())
        self.log.addHandler(ch)

    async def on_partitions_revoked(self, revoked):
        self.log.info(f"Revoking partitions {revoked}, waiting on lock")
        # wait until a single batch processing is done
        async with self.lock:
            pass
        self.log.info("Partitions Revoked")

    async def on_partitions_assigned(self, assigned):
        pass


class Engine:
    """
    Engine used to consume messages from Kafka, using AIOKafka Library.

    Engine has 3 modes:
    - Single Mode: Consume messages one by one
    - Bulk Mode: Consume messages in bulk, using AIOKafka's getmany method
    - Step Mode: Consume messages in bulk, but in steps. Each step will consume messages from a list of topics.

    When you run the bulk mode, the engine will consume messages until :bulk_mode_threshdold: is reached
    on all topics. This is useful when you want to consume messages in bulk, but you don't want to wait
    for all topics to have messages to consume.

    If the bulk_mode_threshold is None, the engine will consume messages until the application is stopped
    or an error occurs.
    """

    __loop: asyncio.AbstractEventLoop

    @final
    def __init__(
        self,
        *handlers: type["BaseHandler"],
        steps: list[list[type["BaseHandler"]]] | None = None,
        bootstrap_servers: str | list[str] = "localhost:9092",
        client_id: str = "paperboy",
        group_id: str = "paperboy",
        auto_offset_reset: str = "earliest",
        sasl_mechanism: str = "PLAIN",
        security_protocol: str = "PLAINTEXT",
        sasl_plain_username: str | None = None,
        sasl_plain_password: str | None = None,
        consumer_bootstrap_servers: str | list[str] | None = None,
        consumer_client_id: str | None = None,
        consumer_auto_offset_reset: str | None = None,
        consumer_sasl_mechanism: str | None = None,
        consumer_security_protocol: str | None = None,
        consumer_sasl_plain_username: str | None = None,
        consumer_sasl_plain_password: str | None = None,
        producer_bootstrap_servers: str | list[str] | None = None,
        producer_client_id: str | None = None,
        producer_sasl_mechanism: str | None = None,
        producer_security_protocol: str | None = None,
        producer_sasl_plain_username: str | None = None,
        producer_sasl_plain_password: str | None = None,
        bulk_mode_timeout_ms: int = 10000,
        bulk_mode_max_records: int = 10000,
        bulk_mode_threshold: int | None = 0,
        commit_interval_sec: int = 10,
        fail_on_exception: bool = False,
        logger_level: int = logging.INFO,
        with_commit: bool = True,
        with_bulk_mode: bool = False,
        with_aiokafka_logs: bool = True,
        healthcheck_host: str = "0.0.0.0",
        healthcheck_port: int = 8080,
    ):
        self.handlers = handlers
        self.steps = steps or []
        self.with_bulk_mode = with_bulk_mode
        self.bulk_mode_timeout_ms = bulk_mode_timeout_ms
        self.bulk_mode_max_records = bulk_mode_max_records
        self.bulk_mode_threshold = bulk_mode_threshold
        self.with_commit = with_commit
        self.fail_on_exception = fail_on_exception
        self.commit_interval_sec = commit_interval_sec
        self.healthcheck_host = healthcheck_host
        self.healthcheck_port = healthcheck_port

        self.__initialize_logger(
            log_level=logger_level,
            with_aiokafka_logs=with_aiokafka_logs,
        )

        self.consumer_options = {
            "bootstrap_servers": consumer_bootstrap_servers or bootstrap_servers,
            "client_id": consumer_client_id or client_id,
            "group_id": group_id,
            "enable_auto_commit": False,
            "auto_offset_reset": consumer_auto_offset_reset or auto_offset_reset,
            "sasl_mechanism": consumer_sasl_mechanism or sasl_mechanism,
            "security_protocol": consumer_security_protocol or security_protocol,
            "sasl_plain_username": consumer_sasl_plain_username or sasl_plain_username,
            "sasl_plain_password": consumer_sasl_plain_password or sasl_plain_password,
        }

        self.producer_options = {
            "bootstrap_servers": producer_bootstrap_servers or bootstrap_servers,
            "client_id": producer_client_id or client_id,
            "sasl_mechanism": producer_sasl_mechanism or sasl_mechanism,
            "security_protocol": producer_security_protocol or security_protocol,
            "sasl_plain_username": producer_sasl_plain_username or sasl_plain_username,
            "sasl_plain_password": producer_sasl_plain_password or sasl_plain_password,
        }

        self.offsets: dict[TopicPartition, int] = {}
        self.tables: set["Table"] = set()

        self.log.info("KafkaConsumerEngine initialized !!")

    @final
    def __configure_consumer(self) -> AIOKafkaConsumer:
        return AIOKafkaConsumer(
            **self.consumer_options,
            ssl_context=create_ssl_context(),
        )

    @final
    def __configure_producer(self) -> AIOKafkaProducer:
        return AIOKafkaProducer(
            **self.producer_options,
            ssl_context=create_ssl_context(),
        )

    @final
    def __initialize_logger(
        self,
        log_level: int = logging.INFO,
        with_aiokafka_logs: bool = True,
    ):
        logging.config.dictConfig(
            {
                "version": 1,
                "handlers": {
                    "aiokafka": {
                        "class": "logging.StreamHandler",
                        "formatter": "aiokafka",
                        "stream": "ext://sys.stdout",
                    },
                },
                "formatters": {
                    "aiokafka": {
                        "format": "\x1b[33;20m%(levelname)s | %(name)s | %(message)s\x1b[0m",
                    },
                },
                "loggers": {
                    "aiokafka": {
                        "handlers": ["aiokafka"] if with_aiokafka_logs else [],
                        "level": log_level,
                        "propagate": True,
                    },
                    "paperboy": {
                        "handlers": [],
                        "level": log_level,
                        "propagate": True,
                    },
                },
            }
        )

        self.log = logging.getLogger(f"paperboy.{self.__class__.__name__}")
        ch = logging.StreamHandler()
        ch.setLevel(log_level)
        ch.setFormatter(PaperboyFormatter())

        self.log.addHandler(ch)

    @final
    async def __bulk_mode(self, handlers: list[type["BaseHandler"]] | list["Table"] | None = None):
        in_bulk = True
        topic_handlers = {handler.topic: handler for handler in handlers or self.handlers}
        topics_in_bulk = {handler.topic: True for handler in handlers or self.handlers}

        self.consumer.subscribe(
            list(topic_handlers.keys()),
            listener=EngineConsumerRebalancer(asyncio.Lock()),
        )

        while in_bulk:
            tasks: list[asyncio.Task] = []

            messages = await self.consumer.getmany(
                timeout_ms=self.bulk_mode_timeout_ms,
                max_records=self.bulk_mode_max_records,
            )

            if not messages:
                break

            consumed_msgs = {tp: len(msgs) for tp, msgs in messages.items()}
            self.log.info(f"Received messages:{consumed_msgs}")
            for tp, msgs in messages.items():
                if tp.topic not in topics_in_bulk.keys():
                    topics_in_bulk[tp.topic] = True

                handler = topic_handlers[tp.topic]
                tasks.append(self.__loop.create_task(handler.handle(msgs)))

                if self.bulk_mode_threshold is not None:
                    highwater = self.consumer.highwater(tp)
                    if highwater <= msgs[-1].offset + self.bulk_mode_threshold:  # type: ignore
                        topics_in_bulk[tp.topic] = False

                self.offsets[tp] = msgs[-1].offset + 1  # type: ignore

            results = await asyncio.wait(tasks, return_when=asyncio.ALL_COMPLETED)
            task_exceptions = [task.exception() for task in results[0]]

            if any(task_exceptions):
                for task_exception in task_exceptions:
                    if task_exception:
                        self.log.exception(task_exception)

                if self.fail_on_exception:
                    self.log.error("An error occured, shutting down...")
                    return
                else:
                    self.log.warning("An error occured, continuing...")

            if self.with_commit:
                self.log.info(f"Committing offsets: {self.offsets}")
                await self.consumer.commit(self.offsets)

            in_bulk = any(topics_in_bulk.values())

        self.consumer.unsubscribe()

    @final
    async def __manual_commit(self):
        self.log.info(f"Starting manual commit loop, with a {self.commit_interval_sec}s interval")

        if not self.with_commit:
            self.log.warning("with_commit is False, cannot start manual commit task")
            return

        # To not restart from the beginning when you move from bulk mode to single mode
        # when the with_commit option is false, we need to set the offsets to the last
        # read message of each topic.
        while True:
            await asyncio.sleep(self.commit_interval_sec)
            self.log.debug("Committing offsets...")
            await self.consumer.commit()

    @final
    async def __single_mode(self):
        self.log.info("Starting Single Mode")

        # Assign the same TP as the bulk mode does before, based on the previous subscription
        if not self.offsets:
            self.consumer.subscribe(
                list(self.topic_handlers.keys()),
                listener=EngineConsumerRebalancer(asyncio.Lock()),
            )
        else:
            self.consumer.assign(list(self.offsets.keys()))

        # Create the manual commit task
        self.__loop.create_task(self.__manual_commit())

        # If the with_commit option is false, we need to set the offsets to the last
        if self.offsets and not self.with_commit:
            self.log.info("with_commit is False, setting offsets to the last read message")
            self.log.debug(f"Offsets: {self.offsets}")
            for tp, offset in self.offsets.items():
                self.log.info(f"Seeking topic {tp} to offset {offset}")
                self.consumer.seek(tp, offset)

        async for msg in self.consumer:
            try:
                await self.topic_handlers[msg.topic].handle(msg)
            except Exception as e:
                self.log.exception(e)
                if self.fail_on_exception:
                    self.log.error("An error occured, shutting down...")
                    return

    @final
    async def __consume(self):
        """
        Launches the consumer, and starts consuming messages from Kafka.
        """
        if len(self.steps) > 0:
            self.log.info(f"Entering Steps Mode with {len(self.steps)} steps")
            for step_index, step in enumerate(self.steps):
                self.log.info(
                    f"Step {step_index + 1}/{len(self.steps)} | "
                    f"Entering Bulk Mode with topics: {[handler.topic for handler in step]}"
                )
                await self.__bulk_mode(handlers=step)

        elif self.with_bulk_mode:
            self.log.info(f"Entering Bulk Mode with topics: {self.topic_handlers.keys()}")
            await self.__bulk_mode()

        await self.__single_mode()

    @final
    async def __initialize_tables(self):
        """
        Initializes all tables registered to the engine.
        """
        if not self.tables:
            return

        self.log.info("Initializing tables...")
        for table in self.tables:
            table.initialize(self.consumer, self.producer)

        self.log.info("Recovering tables")
        await self.__bulk_mode(handlers=list(self.tables))

        # Resetting offsets
        self.offsets = {}

    async def __initialize_signal_handling(self):
        """"""
        signals = (signal.SIGHUP, signal.SIGTERM, signal.SIGINT)
        for s in signals:
            self.__loop.add_signal_handler(s, lambda s=s: asyncio.create_task(self.shutdown(s)))

    @final
    def register_table(self, table: "Table"):
        """
        Registers a table to the engine.

        The table will be initialized when the engine starts.
        """
        if any((table.table_name == reg_table.table_name for reg_table in self.tables)):
            raise ValueError(f"Table {table.table_name} already registered")

        if any((table.changelog_topic_name == reg_table.changelog_topic_name for reg_table in self.tables)):
            raise ValueError(f"Changelog topic {table.changelog_topic_name} already registered")

        if any((table.changelog_topic_name == handler.topic for handler in self.handlers)):
            raise ValueError(f"Changelog topic {table.changelog_topic_name} already handled")

        self.tables.add(table)

    @final
    async def start(self):
        """
        Starts the consumer, and handles the shutdown of the application.

        Use this method with asyncio.run, or in an asyncio loop.
        """
        self.__loop = asyncio.get_running_loop()

        # Starting healthcheck API
        self.__loop.create_task(
            run_api(
                host=self.healthcheck_host,
                port=self.healthcheck_port,
            )
        )

        try:
            self.consumer = self.__configure_consumer()
            self.producer = self.__configure_producer()

            await self.consumer.start()
            if self.producer:
                await self.producer.start()

            if self.consumer._enable_auto_commit:
                self.log.error(
                    "KafkaConsumerEngine handle commiting on its own, "
                    "please set enable_auto_commit=False in the AIOKafkaConsumer constructor. "
                    "If you don't want to commit, please set with_commit=False in the "
                    "Engine constructor."
                )
                await self.shutdown()
                return

            # Initialize Signal Handling
            await self.__initialize_signal_handling()

            # Create Topic Handlers
            self.topic_handlers = {handler.topic: handler for handler in self.handlers}

            # Set producer to all handlers
            for handler in self.handlers:
                handler.set_producer(self.producer)

            # Initialize tables
            await self.__initialize_tables()

            # Consume messages
            await self.__loop.create_task(self.__consume())

        except (ConsumerStoppedError, asyncio.CancelledError):
            pass
        except Exception:
            await self.shutdown()
            return

    @final
    async def shutdown(self, signal: signal.Signals | None = None):
        """
        Stops the consumer, and cancels all async tasks.
        """
        if signal:
            self.log.info(f"Received exit signal {signal.name}...")
        self.log.info("Stopping consumer...")
        await self.consumer.stop()
        self.log.info("Stopping producer...")
        await self.producer.stop()

        self.log.info("Cancelling async tasks...")
        tasks = [t for t in asyncio.all_tasks() if t is not asyncio.current_task()]
        [task.cancel() for task in tasks]

        self.log.info(f"Tasks to cancel: {tasks}")
        await asyncio.gather(*tasks, return_exceptions=True)
        self.log.info("All tasks cancelled, exiting...")

        return
